<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>musicatk_help • musicatk</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="musicatk_help">
<meta property="og:description" content="musicatk">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">musicatk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/musicatk.html">General Workflow</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shiny UI Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/musicatk_help.html">Tutorial</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">References</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/campbio/musicatk/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="musicatk_help_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>musicatk_help</h1>
            
            <h4 class="date">5/16/2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/campbio/musicatk/blob/master/vignettes/musicatk_help.Rmd"><code>vignettes/musicatk_help.Rmd</code></a></small>
      <div class="hidden name"><code>musicatk_help.Rmd</code></div>

    </div>

    
    
<p><img src="image/musicatk_tutorial_workflow.png" width="220"></p>
<div id="import" class="section level2">
<h2 class="hasAnchor">
<a href="#import" class="anchor"></a>Import</h2>
<p>Users can upload their datasets through 1 of 3 different methods. - “Import Files” - upload MAF and VCF files containing the variant data. - “Import TCGA Datasets” - easily import all open access TCGA tumor datasets. - “Import Musica Result Object” - upload existing musica and result objects.</p>
</div>
<div id="import-files" class="section level2">
<h2 class="hasAnchor">
<a href="#import-files" class="anchor"></a>Import Files</h2>
<p>In order to discover or predict signatures, VCF or MAF files need to be uploaded in this section.The files can be browsed through by selecting the ‘browse’ button under the ‘Select Files’ label.</p>
<p><img src="image/import.png" width="1150"></p>
<p>After that, the ‘Add Samples’ button can be pressed to see the added files. At this stage, the delete and undo button next to the file can be pressed to remove it or revert the change respectively. After adding the files, the ‘Import’ button can be pressed to start the importing process.</p>
<p>Once the process is completed, a notification in the bottom right will appear. A variants table from the uploaded files files will also appear which can be downloaded by pressing the ‘Download Variants’ button.</p>
</div>
<div id="import-tcga-datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#import-tcga-datasets" class="anchor"></a>Import TCGA Datasets</h2>
<p>This tab allows selection of single or multiple TCGA datasets directly instead of providing user VCFs and MAFs in the ‘Import Files’ step. This is an optional step and not needed. Users have their own data. The TCGA datasets have been named in the following format: TCGA Abbreviation - Full TCGA tumor name. The original TCGA page for the tumor name list can be referred to by selecting the ‘Full Tumor List’ link right next to the ‘Import’ button. Once the datasets have been selected, the ‘Import’ button at the bottom can be pressed. Once finished, a completion notification will appear at the bottom right. <img src="image/TCGA.png" width="1212"> ## Import Musica This tab can be used to import user’s musica result or musica objects in .rda or .rds format which will allow for direct downstream analysis in the workflow. The browse button in the musica result or musica result object can be selected for selecting files. Once selected, a bar saying file upload is completed will appear. By default, the musica result objects will be named after the file’s name but it can be changed in the text box labeled “Name your musica result object”. Once finished, a completion notification message in the bottom right and a variants table of the uploaded file will appear. At this stage, the next steps in the workflow can be accessed directly. If a musica result object has been uploaded, all the downstream tabs are now accessible and if a musica object has been uploaded then the users can skip the Create Musica tab and move to the build tables tab directly. The variants table can be downloaded using the "Download Variants’ button.</p>
</div>
<div id="create-musica-object" class="section level2">
<h2 class="hasAnchor">
<a href="#create-musica-object" class="anchor"></a>Create Musica Object</h2>
<p>This tab can be used to create a musica object for discovery and prediction of signatures from the given variants. The reference genome can be selected from the drop down menu labeled “Choose Genome” that matches the genome build of the variants and the ‘Create Musica Object’ button can then be pressed. Once it’s done, a completion notification message at the bottom right and a variant summary table will appear. The musica object can be downloaded by pressing the ‘Download Musica Object’ button.</p>
<p><img src="image/create_musica.png" width="1086"><img src="image/create_musica1.png" width="575"></p>
</div>
<div id="annotations" class="section level2">
<h2 class="hasAnchor">
<a href="#annotations" class="anchor"></a>Annotations</h2>
<p>Sample annotations can be used to store information about each sample such as tumor type or treatment status. They are optional and can be used in downstream plotting functions such as plot_exposures or plot_umap to group or color samples by a particular annotation.</p>
<p><img src="image/AnnotationsTab.png" width="1377"><br> Annotations are provided in a character delimited text file which includes a sample name column. The sample names must match the sample names in the selected musica object. NA values will be used for any samples not present in the annotation file.</p>
<p><img src="image/AnnotationOptions.png" width="1142"><br> After selecting the correct delimiter, a data table will appear below to show the annotations that will be added to the musica object. If your annotation file does not contain a header, deselect the “Header” radio button. Choose the column that contains the sample names from the “Sample Name Columns” dropdown and then click “Add Annotation”. <br><img src="image/AnnotationTable.png" width="1087"></p>
</div>
<div id="build-tables" class="section level2">
<h2 class="hasAnchor">
<a href="#build-tables" class="anchor"></a>Build Tables</h2>
<p>The “Build Tables” tab is used to generate count tables for different mutation type schemas which are the input to mutational signature discovery or prediction functions. To build the count tables, the user must select 1 of the 5 standard motifs in the “Select Count Table” dropdown. SBS96 - Motifs are the six possible single base pair mutation types times the four possibilities each for upstream and downstream context bases (464 = 96 motifs) <br> SBS192_Trans - Motifs are an extension of SBS96 multiplied by the transcriptional strand (translated/untranslated), can be specified with “Transcript_Strand”. <br> SBS192_Rep - Motifs are an extension of SBS96 multiplied by the replication strand (leading/lagging), can be specified with “Replication_Strand”. <br> DBS - Motifs are the 78 possible double-base-pair substitutions. <br> INDEL - Motifs are 83 categories intended to capture different categories of indels based on base-pair change, repeats, or microhomology, insertion or deletion, and length. In addition to selecting a motif, the user must also provide the reference genome.</p>
<p><img src="image/TableTab.png" width="823"></p>
</div>
<div id="signatures-and-exposures" class="section level2">
<h2 class="hasAnchor">
<a href="#signatures-and-exposures" class="anchor"></a>Signatures and Exposures</h2>
<p>The user has the option of discovering novel signatures and exposures, or using previously discovered signatures to predict exposures. To discover signatures use the “Discover” tab. To predict signature exposures, use the “Predict” tab.</p>
<div id="discover" class="section level3">
<h3 class="hasAnchor">
<a href="#discover" class="anchor"></a>Discover</h3>
<p>Mutational signatures and exposures can be discovered using methods such as Latent Dirichlet Allocation (LDA) or Non-Negative Matrix Factorization (NMF). These algorithms will deconvolute a matrix of counts for mutation types in each sample to two matrices: 1) a “signature” matrix containing the probability of each mutation type in each sample and 2) an “exposure” matrix containing the estimated counts for each signature in each sample. Before mutational discovery can be performed, variants from samples first need to be stored in a musica object using the create_musica function and mutation count tables need to be created using functions such as build_standard_table.</p>
<p>You can select any count table for signature discovery. To obtain biologically significant results, it is important to select a reasonable number of expected signatures. In this tutorial, we chose 8 signatures.</p>
<p><img src="image/DiscoverTab.png" width="833"></p>
</div>
<div id="predict" class="section level3">
<h3 class="hasAnchor">
<a href="#predict" class="anchor"></a>Predict</h3>
<p>Exposures for samples will be predicted using an existing set of signatures stored in a musica_result object. Algorithms available for prediction include “lda”, “decompTumor2Sig”, and “deconstructSigs”.</p>
<p><img src="image/PredictTab.png" width="825"></p>
<p>The “Signatures to Predict” dropdown contains all the signatures in the result object selected from the “Result to Predict” dropdown. You can search this dropdown and select multiple signatures. In our example we use the “lda” algorithm to predict exposures of 6 Cosmic signatures in our samples.</p>
<p><img src="image/PredictDropdown.png" width="546"></p>
</div>
</div>
<div id="data-visualization" class="section level2">
<h2 class="hasAnchor">
<a href="#data-visualization" class="anchor"></a>Data Visualization</h2>
<p>The data visualization tab can make customized plots for signatures and exposures predicted from LDA and NMF algorithms using the ggplot2 package. An option of making interactive plots using plotly is also provided.</p>
<div id="signatures" class="section level3">
<h3 class="hasAnchor">
<a href="#signatures" class="anchor"></a>Signatures</h3>
<p>The signature plot is presented only using bar plot with each bar representing the probability of each type of mutation. By default, signatures are named by numbers, but an option of renaming signatures is provided if you want to name them otherwise, such as the possible etiology.</p>
<p><img src="image/visual_sig.png" width="936"></p>
<p>In this tutorial, we found 8 single-base signatures from the mixture of lung adenocarcinoma, lung squamous cell carcinoma, and skin cutaneous melanoma samples.</p>
<p><img src="image/sig_plot.png" width="1538"></p>
</div>
<div id="exposures" class="section level3">
<h3 class="hasAnchor">
<a href="#exposures" class="anchor"></a>Exposures</h3>
<p>To visualize the exposure of each signature for each sample, we provided three options, including bar plot, box plot, and violin plot.</p>
<p><img src="image/visual_exp.png" width="968"></p>
<p>By default, a stacked bar plot sorted by the total number of mutations is used. Each stacked bar shows the proportion of exposure of each signature.</p>
<p><img src="image/exp_plot1.png" width="984"></p>
<p>The stacked bar plot can be ordered by signatures. If <code>Signatures</code> is selected in <code>Sort By</code> option, a bucket list will show up to allow you select the signatures you want to use by dragging them from the left box to the right box. Users can also set limit on the number of samples to display.</p>
<p><img src="image/sorting.png" width="890"></p>
<p>This stacked bar plot is now ordered by the exposure of signature 1 and only top 400 samples were included here.</p>
<p><img src="image/exp_plot2.png" width="892"></p>
<p>Box plot and violin plot can be used to visualize the distribution of exposures or compare exposures between different groups of samples.</p>
<p><img src="image/exp_box_setting1.png" width="886"></p>
<p>By default, a box plot of exposure for each signature will be shown.</p>
<p><img src="image/exp_plot3.png" width="892"></p>
<p>If an annotation file is provided, then you can visually compare the exposure of signatures among different groups.</p>
<p><img src="image/exp_box_setting2.png" width="891"></p>
<p>In this box plot, exposures of each signature were grouped by tumor types. We can find that signatures 1 and 5 were highly exposed in lung cancer samples, while signatures 3 and 8 were enriched in skin cancer samples.</p>
<p><img src="image/exp_plot4.png" width="892"></p>
<p>We can also group samples by signatures and then color by tumor types.</p>
<p><img src="image/exp_box_setting3.png" width="888"></p>
<p>This plot can let you directly compare how each signatures were differentially exposed among three tumor types.</p>
<p><img src="image/exp_plot5.png" width="892"></p>
</div>
</div>
<div id="additional-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#additional-analysis" class="anchor"></a>Additional Analysis</h2>
<div id="compare-signatures" class="section level3">
<h3 class="hasAnchor">
<a href="#compare-signatures" class="anchor"></a>Compare Signatures</h3>
<p>Compare two result objects to find similar signatures. The threshold acts as a cutoff similarity score and can be any value between 0 and 1. Results will populate in a data table below which can be downloaded.</p>
<p><img src="image/CompareTab.png" width="1383"></p>
<p>Using the cosine metric with a threshold of 0.8 we can identify 7 pairs of similar signatures. Interestingly, we identified a pair with a high cosine similarity score - signature1 and Cosmic SBS4. Our results indicate we may have identified a tobacco smoke signature among our tumor samples. This should not be a surprise given many of our samples are from lung tumors.</p>
<p><img src="image/CompareTable.png" width="735"></p>
</div>
<div id="exposure-differential-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#exposure-differential-analysis" class="anchor"></a>Exposure Differential Analysis</h3>
<p>The “Exposure Differential Analysis” tab is used to run differential analysis on the signature exposures of annotated samples. There are 3 methods to perform the differential analysis: Wilcoxon Rank Sum Test, Kruskal-Wallis Rank Sum Test, and a negative binomial regression (glm).</p>
<p><img src="image/DATab.png" width="1373"></p>
<p>When using the Wilcoxon Rank Sum Test, any two groups will be compared in a pairwise fashion. Any isolated group will be ignored. Below we display the Wilcoxon Rank Sum Test results between LUAD and SKCM. Note that LUAD was ignored since it has no pair.</p>
<p><img src="image/DAOptions.png" width="1161"></p>
<p>From our table, we can see several of our signatures are differentially exposed across tumor types. From our comparison analysis, we can hypothesize signature1 may be related to tobacco smoke. We may hypothesize tobacco smoke is a primary driving mechanism for lung cancer development, but not for skin cancer.</p>
<p><img src="image/WilcoxonTable.png" width="1150"></p>
</div>
<div id="clustering" class="section level3">
<h3 class="hasAnchor">
<a href="#clustering" class="anchor"></a>Clustering</h3>
<p>The clustering subtab provides several algorithms to cluster samples based on exposure of each signature. After selecting the musica result object, it is recommended to use <code>Explore Number of Clusters</code> box to find the optimal number of clusters in your data. Different clustering algorithms and three metrics including within cluster sum of squares, averaged silouette coefficient, and gap statistics are provided here for exploration. All algorithms were imported from <a href="https://cran.r-project.org/web/packages/factoextra/index.html">factoextra</a> and <a href="https://cran.r-project.org/web/packages/cluster/index.html">cluster</a>. (Note: If gap statistic is selected, it will take much longer time to generate the plot.)</p>
<p><img src="image/cluster1.png" width="863"></p>
<p>This is a connected scatter plot shows the within cluster sum of squares for each number of clusters predicted using hierarchical clustering. The “elbow” method can be used to determine the optimal number of clusters.</p>
<p><img src="image/wss.png" width="553"></p>
<p>The <code>Clustering</code> box is where you perform the clustering analysis. In addition to clustering algorithm, several methods for calculating dissimilarity matrix, imported from <a href="https://cran.r-project.org/web/packages/philentropy/index.html">philentropy</a> package, are also provided.</p>
<p><img src="image/cluster2.png" width="574"></p>
<p>This table is the output of clustering analysis, combined with annotation.</p>
<p><img src="image/cluster_tbl.png" width="574"></p>
<p>In the <code>Visulaization</code> box, users can make scatter plots to visualize the clustering results on a UMAP panel, calculated from signature exposures. Three types of plots are provided.</p>
<p><img src="image/cluster_visual.png" width="884"></p>
<p>If <code>Signature</code> is selected, samples are grouped by clusters and multiplicated by the number of signatures. For each column, samples are colored by exposure of a signature.</p>
<p><img src="image/umap1.png" width="553"></p>
<p>If <code>Annotation</code> is selected, an additional select box will show up and let you choose one type of annotation of interest. Then, you can make a plot grouping samples by both clusters and annotation.</p>
<p><img src="image/umap2.png" width="553"></p>
<p>If <code>None</code> is selected, a single scatter plot, colored by clusters, will be made.</p>
<p><img src="image/umap3.png" width="553"></p>
</div>
<div id="heatmap" class="section level3">
<h3 class="hasAnchor">
<a href="#heatmap" class="anchor"></a>Heatmap</h3>
<p>Please select the result object you want to use for heatmap visualization from the dropdown menu in the start labeled ‘Select Result’. Choose from different settings and press the ‘Plot’ button to see your heatmap. If you want to normalize your data, check the ‘Proportional’ setting and for z-scale normalization, check the ‘z-scale’ option. You can also see column or row names by checking ‘Show column names’ and ‘Show row names’ resepctively. If you want to subset by signatures, press ‘Selected Signatures’ to select the ones you want. You can also select the ‘Samples’ and ‘Annotation’ option to subset by the available samples and annotations respectively. <img src="image/heatmap.png" width="1054"><img src="image/heatmap1.png" width="666"></p>
</div>
</div>
<div id="download" class="section level2">
<h2 class="hasAnchor">
<a href="#download" class="anchor"></a>Download</h2>
<p>You can download your musica object and musica result object from their labeled respective drop down menus as .rds files. Please select the name of the musica result or musica object that you want to download from their drop down menus and press the respective download button. include_graphics(paste0(PATH,“s/download.png”)</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Aaron Chevalier, Joshua D. Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
