<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Builds count tables using various mutation type schemas — build_standard_table • musicatk</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Builds count tables using various mutation type schemas — build_standard_table" />
<meta property="og:description" content="Generates count tables for different mutation type schemas which can be 
used as input to the mutational signature discovery or prediction functions.
&quot;SBS96&quot; generates a table for single base substitutions following the
standard 96 mutation types derived from the trinucleotide context.
&quot;SBS192&quot; is the 96 mutation type schema with the addition of
transcriptional strand or replication strand information added to each base.
&quot;DBS&quot; generates a table for the double base substitution schema 
used in COSMIC V3. &quot;Indel&quot; generates a table for insertions and 
deletions following the schema used in COSMIC V3." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">musicatk</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.6.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="https://www.camplab.net/musicatk">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/articles/tutorial_tcga_cnsl.html">Analysis in the R console</a>
    </li>
    <li>
      <a href="../articles/articles/tutorial_tcga_ui.html">Analysis with the user interface</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/campbio/musicatk">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Builds count tables using various mutation type schemas</h1>
    <small class="dont-index">Source: <a href='https://github.com/campbio/musicatk/blob/master/R/standard_tables.R'><code>R/standard_tables.R</code></a></small>
    <div class="hidden name"><code>build_standard_table.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generates count tables for different mutation type schemas which can be 
used as input to the mutational signature discovery or prediction functions.
<code>"SBS96"</code> generates a table for single base substitutions following the
standard 96 mutation types derived from the trinucleotide context.
<code>"SBS192"</code> is the 96 mutation type schema with the addition of
transcriptional strand or replication strand information added to each base.
<code>"DBS"</code> generates a table for the double base substitution schema 
used in COSMIC V3. <code>"Indel"</code> generates a table for insertions and 
deletions following the schema used in COSMIC V3.</p>
    </div>

    <pre class="usage"><span class='fu'>build_standard_table</span><span class='op'>(</span>
  <span class='va'>musica</span>,
  <span class='va'>g</span>,
  <span class='va'>table_name</span>,
  strand_type <span class='op'>=</span> <span class='cn'>NULL</span>,
  overwrite <span class='op'>=</span> <span class='cn'>FALSE</span>,
  verbose <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>musica</th>
      <td><p>A <code><a href='musica-class.html'>musica</a></code> object.</p></td>
    </tr>
    <tr>
      <th>g</th>
      <td><p>A BSgenome object indicating which genome
reference the variants and their coordinates were derived from.</p></td>
    </tr>
    <tr>
      <th>table_name</th>
      <td><p>Name of standard table to build. One of <code>"SBS96"</code>,
<code>"SBS192"</code>, <code>"DBS"</code>, or <code>"Indel"</code>.</p></td>
    </tr>
    <tr>
      <th>strand_type</th>
      <td><p>Strand type to use in SBS192 schema. One of
<code>"Transcript_Strand"</code> or <code>"Replication_Strand"</code>.
Only used if <code>table_name = SBS192</code>.</p></td>
    </tr>
    <tr>
      <th>overwrite</th>
      <td><p>If <code>TRUE</code>, any existing count table with the same
name will be overwritten. If <code>FALSE</code>, then an error will be thrown
if a table with the same name exists within the <code>musica</code> object.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Show progress bar for processed samples</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>No object will be returned. The count tables will be automatically
added to the <code>musica</code> object.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='r-in'><span class='va'>g</span> <span class='op'>&lt;-</span> <span class='fu'><a href='select_genome.html'>select_genome</a></span><span class='op'>(</span><span class='st'>"19"</span><span class='op'>)</span></span>
<span class='r-in'></span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>musica</span><span class='op'>)</span></span>
<span class='r-in'><span class='fu'>build_standard_table</span><span class='op'>(</span><span class='va'>musica</span>, <span class='va'>g</span>, <span class='st'>"SBS96"</span>, overwrite <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Building count table from SBS with SBS96 schema</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>Overwriting counts table: SBS96</span>
<span class='r-in'></span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>musica</span><span class='op'>)</span></span>
<span class='r-in'><span class='fu'><a href='annotate_transcript_strand.html'>annotate_transcript_strand</a></span><span class='op'>(</span><span class='va'>musica</span>, <span class='st'>"19"</span><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   403 genes were dropped because they have exons located on both strands</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   of the same reference sequence or on more than one reference sequence,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   so cannot be represented by a single genomic range.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Use 'single.strand.genes.only=FALSE' to get all the genes in a</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   GRangesList object, or use suppressMessages() to suppress this message.</span>
<span class='r-in'><span class='fu'>build_standard_table</span><span class='op'>(</span><span class='va'>musica</span>, <span class='va'>g</span>, <span class='st'>"SBS192"</span>, <span class='st'>"Transcript_Strand"</span><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Building count table from SBS and Transcript_Strand with SBS192 schema</span>
<span class='r-in'></span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>musica</span><span class='op'>)</span></span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>rep_range</span><span class='op'>)</span></span>
<span class='r-in'><span class='fu'><a href='annotate_replication_strand.html'>annotate_replication_strand</a></span><span class='op'>(</span><span class='va'>musica</span>, <span class='va'>rep_range</span><span class='op'>)</span></span>
<span class='r-in'><span class='fu'>build_standard_table</span><span class='op'>(</span><span class='va'>musica</span>, <span class='va'>g</span>, <span class='st'>"SBS192"</span>, <span class='st'>"Replication_Strand"</span><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Building count table from SBS and Replication_Strand with SBS192 schema</span>
<span class='r-in'></span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>dbs_musica</span><span class='op'>)</span></span>
<span class='r-in'><span class='fu'>build_standard_table</span><span class='op'>(</span><span class='va'>dbs_musica</span>, <span class='va'>g</span>, <span class='st'>"DBS"</span>, overwrite <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Building count table from DBS with DBS78 schema</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>Overwriting counts table: DBS78</span>
<span class='r-in'></span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>indel_musica</span><span class='op'>)</span></span>
<span class='r-in'><span class='fu'>build_standard_table</span><span class='op'>(</span><span class='va'>indel_musica</span>, <span class='va'>g</span>, table_name <span class='op'>=</span> <span class='st'>"INDEL"</span><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Building count table from INDELs with IND83 schema</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>data length differs from size of matrix: [43990 != 83 x 5]</span>
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Aaron Chevalier, Joshua D. Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


